core/trino-main/src/main/java/io/trino/security/AccessControlManager.java

```diff
@@ -1248,26 +1249,30 @@ public List<ViewExpression> getRowFilters(SecurityContext context, QualifiedObje
     }
 
     @Override
-    public List<ViewExpression> getColumnMasks(SecurityContext context, QualifiedObjectName tableName, String columnName, Type type)
+    public Optional<ViewExpression> getColumnMask(SecurityContext context, QualifiedObjectName tableName, String columnName, Type type)
     {
         requireNonNull(context, "context is null");
         requireNonNull(tableName, "tableName is null");
 
         ImmutableList.Builder<ViewExpression> masks = ImmutableList.builder();
 
-        // connector-provided masks take precedence over global masks
         ConnectorAccessControl connectorAccessControl = getConnectorAccessControl(context.getTransactionId(), tableName.getCatalogName());
         if (connectorAccessControl != null) {
-            connectorAccessControl.getColumnMasks(toConnectorSecurityContext(tableName.getCatalogName(), context), tableName.asSchemaTableName(), columnName, type)
-                    .forEach(masks::add);
+            connectorAccessControl.getColumnMask(toConnectorSecurityContext(tableName.getCatalogName(), context), tableName.asSchemaTableName(), columnName, type)
+                    .ifPresent(masks::add);
         }
 
         for (SystemAccessControl systemAccessControl : getSystemAccessControls()) {
-            systemAccessControl.getColumnMasks(context.toSystemSecurityContext(), tableName.asCatalogSchemaTableName(), columnName, type)
-                    .forEach(masks::add);
+            systemAccessControl.getColumnMask(context.toSystemSecurityContext(), tableName.asCatalogSchemaTableName(), columnName, type)
```

Comment:
If we will stick with having only one mask altogether no matter where it comes from, then the connector mask should be dropped here if any and give place to the system mask. This behavior is guaranteed in the doc: https://trino.io/docs/current/security/built-in-system-access-control.html
`A system access control enforces authorization at a global level, before any connector level authorization`
We may also want to update the doc to mention what will happen if masks for the same column are defined in both system and connector AC.

Reply:
That's an interesting point. In the current implementation, the masks stack on top of each other, with system level ones having precedence. As we've been discussing, stacking masks is problematic, so that leaves us with two options:

1. System-level column mask takes precedence and connector-level mask is ignored if one is also provided.
2. Fail the query if a system-level and a connector-level mask exist.

I'm not sure yet which option is best. The desire for `A system access control enforces authorization at a global level, before any connector level authorization` doesn't quite apply here, since the masks don't behave like regular permission checks.

The issue with option 1 is that it may lead to inadvertent security holes if someone installs a system level plugin after the fact that provides a weaker mask that the connector-level access control did.

@dain, do you have any thoughts?

Reply:
> weaker mask 

I don't think mask can be anyhow ordered. 

If we go with option 2 then it is possible to relax it in the future to option 1. 

The fact we had a list of masks was because it was addressing a corner case which should be very rare. Integrating with multiple authorization systems at once is complex. If you consider that each of them can provide masking for the same columns is a madness :)

Let's disallow this and wait to see if such use case is even real.